chore: add loan explanation (#382)
Agoric/documentation: main/zoe/guide/contracts/loan.md

# Loan

<Zoe-Version/>

##### [View the code on Github](https://github.com/Agoric/agoric-sdk/tree/master/packages/zoe/src/contracts/loan)
##### [View all contracts on Github](https://github.com/Agoric/agoric-sdk/tree/master/packages/zoe/src/contracts)

> The basic loan contract allows a  borrower to add collateral of a

Reviewer: ```suggestion
The basic loan contract lets a  borrower add collateral of a
```

particular brand and get a loan of another brand. The collateral (also
known as margin) must be greater than the loan value, at an amount set
by the Maintenance Margin Requirement (`mmr`) in the terms of the

> contract.

Reviewer: This is a little confusing/unclear to me. I know that in stocks and similar, the margin requirement is usually set as a percentage, but this reads more like it's a set amount. I'd suggest rephrasing as whichever is appropriate:
"The collateral (also called *margin*) must be greater than the loan value. How much greater is specified by the
Maintenance Margin Requirement (`mmr`) in the contract's terms. The `mmr` value is an absolute value added to the loan value. For example, if the `mmr` is 1,000, and the loan value is 500, the collateral must be greater than 1500."

Or replace the last with "The `mmr` value is expressed as a minimum percentage of the collateral with respect to the loan value. So if the `mmr` is 25 and the loan value is 100, the collateral must be greater than 100 + (.25 x 100) or 125."   

Yes, there's at least one more way of doing this, depending on just what/how the `mmr` is expressed. My point is that as written, it could use more detail and an example as to how this contract is associating/using `mmr`, collateral, and loan value.

Author: Gotcha, ok thanks! Will revise

The loan does not have a distinct end time. Rather, if the
value of the collateral changes such that insufficient margin is
provided, the collateral is liquidated, and the loan is closed. At any

> time, the borrower can add collateral or repay the loan with interest,

Reviewer: Might add a parenthetical about where the interest rate and compounding period are specified.

Author: Hmm, I think that is explained enough below

closing the loan. The borrower can set up their own margin calls by
getting the [`priceAuthority`](/zoe/guide/price-authority.md) from the terms and calling

> `E(priceAuthority).quoteWhenLT(allCollateral, x)` where x is the value

Reviewer 2: ```suggestion
`E(priceAuthority).quoteWhenLT(allCollateralAmount, x)` where x is the value
```

of the collateral in the Loan brand at which they want a reminder to
addCollateral.

Note that all collateral must be of the same brand and all of the
loaned amount and interest must be of the same (separate) brand.

## Terms
* `mmr` (default = 150) - the Maintenance Margin Requirement, in
   percent. The default is 150, meaning that collateral should be
   worth at least 150% of the loan. If the value of the collateral

>    drops below `mmr`, liquidation occurs.

Reviewer 2: It's not immediately triggered, so I want to say it "can" occur.

```suggestion
   drops below `mmr`, liquidation can occur.
```

>    drops below `mmr`, liquidation occurs.

Reviewer: ```suggestion
   drops below the `mmr` specified percentage of the loan, liquidation occurs.
```

> * [`priceAuthority`](/zoe/guide/price-authority.md) - will be used for getting the current value of

Reviewer: ```suggestion
* [`priceAuthority`](/zoe/guide/price-authority.md) - used for getting the current value of
```

   collateral and setting liquidation triggers.
* `autoswapInstance` - The running contract instance for an
   [Autoswap](./autoswap.md) or [Multipool

>    Autoswap](./multipoolAutoswap.md) installation. The `publicFacet`
>    of the instance is used for producing an invitation to sell the
>    collateral on liquidation.

Reviewer: ```suggestion
   Autoswap](./multipoolAutoswap.md) installation. The instance's `publicFacet`
   produces an invitation to sell the collateral on liquidation.
```

Author: This sounds like it is active, but the passive voice is more appropriate because we are calling a method on the publicFacet.

Author: Revised

* `periodAsyncIterable` - the [asyncIterable](https://javascript.info/async-iterators-generators) used for notifications
   that a period has passed, on which compound interest will be
   calculated using the `interestRate`.
* `interestRate` - the rate in [basis points](https://www.investopedia.com/terms/b/basispoint.asp) that will be multiplied
   with the debt on every period to compound interest.

## IssuerKeywordRecord

> The following is used for all keyword records, regardless of role in

Reviewer: ```suggestion
All keyword records use the following, regardless of their role in
```

the contract:

* Keyword: 'Collateral' - The issuer/payment for the digital assets to be
   escrowed as collateral.
* Keyword: 'Loan' - The issuer/payment for the digital assets to be loaned
   out.

## The Lender

The lender puts up the amount to be loaned to the borrower, but has no
further actions. The loan is ongoing until it is paid back entirely or
liquidated, at which point the lender receives a payout. This means

> that the payout for the lender will be in Loan-branded digital assets,
> not Collateral-brand. (The only exception to this is if the scheduling

Reviewer: ```suggestion
the lender's payout will be in Loan-branded digital assets,
not Collateral-brand. (The only exception is if the scheduling
```

of liquidation triggers with the `priceAuthority` results in a error. In

> that case, we have no choice but to give the collateral to the lender.

Reviewer: ```suggestion
that case, we must give the collateral to the lender.
```

The borrower has already exited with their loan.)

> The lender will want the interest earned from the loan + their refund
> or the results of the liquidation. If the price of collateral drops
> before we get the chance to liquidate, the total payout could be zero.

Reviewer: ```suggestion
The lender will want the loan interest in addition to either their refund
or the liquidation results. If the collateral price drops
before we get the chance to liquidate, the total payout could be zero.
```

Reviewer: In the above, instead of "we get the chance to liquidate", should it be "they get the chance to liquidate"? Not quite sure who's doing the liquidating here.

Author: the contract is doing the liquidating

Therefore, the lender cannot `want` anything in their proposal.

The lender must be able to exit on demand until borrowing occurs. If the exit rule was

> `waived`, a borrower would be able to hold on to their invitation and

Reviewer: ```suggestion
`waived`, a borrower could hold on to their invitation and
```

> effectively lock up the lender's Loan forever.  When the borrowing

Reviewer: Not sure about "Loan"; you've been using "loan" for a while now.

Author: I think I mean Loan-branded tokens but @Chris-Hibbert  didn't want me to use the word "tokens"

Author: Revised

occurs, the collateral is moved to a special collateral seat to

> prevent the lender from being able to exit with collateral before the

Reviewer: ```suggestion
prevent the lender from exiting with collateral before the
```

> contract ends through repayment or liquidation.

Reviewer: ```suggestion
contract ends due to repayment or liquidation.
```

## The Borrower

The borrower receives an invitation to join, makes an offer with Zoe
escrowing their collateral, and receives their loan. The borrower seat
is exited at this point so the borrower gets the payout of their loan,

> but the borrower also gets an object (`borrowFacet`) as their `offerResult` that will
> let them continue to interact with the contract.

Reviewer: ```suggestion
but the borrower also gets an object (`borrowFacet`) as their `offerResult` that
lets them continue interacting with the contract.
```

> Once the loan has started, the borrower can repay the loan in its

Reviewer: ```suggestion
Once the loan starts, the borrower can repay the loan in its
```

> entirety (at which the lender receives the loan amount back plus
> interest and the contract closes), or add more collateral to stave off
> liquidation.

Reviewer: ```suggestion
entirety at any time (at which point the lender receives the loan amount back plus
interest, and the contract closes), or add more collateral to prevent
liquidation.
```

## Contract Shutdown

> The contract will shutdown under 3 conditions:

Reviewer: ```suggestion
The contract shuts down under 3 conditions:
```

> 1. The loan (plus interest) is repaid.
>    * In this case the lender gets the repayment and the borrower gets
>     their collateral back.
> 2. The value of the collateral drops and the collateral must be
>    liquidated.
>    * In this case the lender gets the outcome of the sale of the
>      collateral, and the borrower still keeps their loan.
> 3. An error occurs when trying to use the priceAuthority.
>    * In this case the lender gets the collateral, and the borrower
>      keeps their loan.

Reviewer: ```suggestion
1. The loan (plus interest) is repaid.
   * The lender gets the repayment and the borrower gets
    their collateral back.
2. The value of the collateral drops and the collateral must be
   liquidated.
   * The lender gets the outcome of the collateral sale, and the borrower keeps their loan.
3. An error occurs when trying to use the priceAuthority.
   * The lender gets the collateral, and the borrower keeps their loan.
```

## Debt and Interest Calculation 

Interest is calculated and compounded when the
`periodAsyncIterable` pushes a new value. The interest rate per period
is defined by the `interestRate` parameter.

## Scheduling Liquidation

Liquidation is scheduled using the `priceAuthority` parameter.
Specifically, the contract gets a promise resolved when the value of the

> collateral falls below a particular trigger value defined by the `mmr`
> parameter: 

Reviewer: ```suggestion
collateral falls below a trigger value defined by the `mmr` parameter: 
```

```js
  const internalLiquidationPromise = E(priceAuthority).quoteWhenLT(
    allCollateral,
    liquidationTriggerValue,
  );
  internalLiquidationPromise.then(liquidate);
```

> The borrower can set up their own margin calls to be forewarned of a
> potential liquidation by

Reviewer: ```suggestion
The borrower can self-forewarn about a potential liquidation by setting up their own margin calls.
```

> getting the [`priceAuthority`](/zoe/guide/price-authority.md) from the terms and calling

Reviewer: ```suggestion
They do this by getting the [`priceAuthority`](/zoe/guide/price-authority.md) from the terms and calling
```

> `E(priceAuthority).quoteWhenLT(allCollateral, x)` where x is the value

Reviewer 2: This would benefit from a code sample showing how to use `.then()` to invoke the caller's handler.

Author: Yup!

of the collateral in the Loan brand at which they want a reminder to
addCollateral.

## Liquidating

Actual liquidation is done through an Autoswap or Multipool Autoswap
instance, regardless of the current price within the Autoswap
instance. Even if the price is worse or better than what our `priceAuthority`
quoted, we still liquidate.

> ## Future directions

Reviewer 2: Add ability for the lender to withdraw accrued interest. Otherwise, they've written a perpetual loan that they might never benefit from.

Author: Ah, they can't though, unless the repayment is happening partially as well. The interest is just debt not represented by anything escrowed.
